pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,345 @@
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type DrainOptions struct {
+	client             *client.Client
+	factory            *cmdutil.Factory
+	Force              bool
+	GracePeriodSeconds int
+	mapper             meta.RESTMapper
+	nodeInfo           *resource.Info
+	out                io.Writer
+	typer              runtime.ObjectTyper
+}
+
+const (
+	lame_long = `Mark nodes as unschedulable.
+`
+	lame_example = `# Mark node "foo" as unschedulable.
+$ kubectl lame foo
+`
+)
+
+func NewCmdLame(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	return &cobra.Command{
+		Use:     "lame NODE",
+		Short:   "Mark node as unschedulable",
+		Long:    lame_long,
+		Example: lame_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunLame(true))
+		},
+	}
+}
+
+const (
+	unlame_long = `Mark nodes as schedulable.
+`
+	unlame_example = `# Mark node "foo" as schedulable.
+$ kubectl unlame foo
+`
+)
+
+func NewCmdUnlame(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	return &cobra.Command{
+		Use:     "unlame NODE",
+		Short:   "Mark node as schedulable",
+		Long:    unlame_long,
+		Example: unlame_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunLame(false))
+		},
+	}
+}
+
+const (
+	drain_long = `Drain node in preparation for maintenance.
+
+The given node will be marked unschedulable to prevent new pods from arriving.
+Then all running pods will be removed from the machine.  If there are any
+unreplicated pods, no pods will be touched without --force.
+
+When you are ready to put the node back into service, use kubectl unlame.
+`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunDrain(cmd))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.")
```

Comment:
We are currently using -1 as a magic value that means "don't specify a
grace period", and I can't see how to do that directly with Duration().
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/delete.go#L98

On Tue, Nov 17, 2015 at 3:15 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/drain.go
> <https://github.com/kubernetes/kubernetes/pull/16698#discussion_r45137381>
> :
>
> > +
> > +func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
> > +	options := &DrainOptions{factory: f, out: out}
> > +
> > +	cmd := &cobra.Command{
> > +		Use:     "drain NODE",
> > +		Short:   "Drain nodes in preparation for maintenance",
> > +		Long:    drain_long,
> > +		Example: drain_example,
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			cmdutil.CheckErr(options.SetupDrain(cmd, args))
> > +			cmdutil.CheckErr(options.RunDrain(cmd))
> > +		},
> > +	}
> > +	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
> > +	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.")
>
> For kubectl, I prefer duration:
>
> https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/rollingupdate.go#L92
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16698/files#r45137381>.
>